lib/shopify_cli/theme/dev_server/hot-reload-tae.js

```diff
@@ -0,0 +1,173 @@
+    const namespace = window.__SHOPIFY_CLI_ENV__;
+    return namespace.mode;
+  }
+
+  function querySelectHotReloadElements(handle) {
+    // Gets all blocks (app and embed) with specified handle TODO: add app ID check here
+    const blocks = Array.from(document.querySelectorAll(`[data-block-handle$='${handle}']`));
+    if (blocks.length){
+      const queryString = 'shopify-section-template';
+      const is_section = blocks[0].closest(`[id^=${queryString}]`) !== null;
+      if (is_section) return [blocks.map((block) => block.closest(`[id^=${queryString}]`)), APP_BLOCK];
+
+      return [blocks, APP_EMBED_BLOCK];
+    }
+    return [[], null];
+  }
+
+  function isFullPageReloadMode(){
+    return reloadMode() === 'full-page';
+  }
+
+  function isReloadModeActive(){
+    return reloadMode() !== 'off';
+  }
+
+  function isRefreshRequired(files) {
+    if (isFullPageReloadMode()) {
+      return true;
+    }
+    return files.some((file) => !isCssFile(file) && !isBlockFile(file));
+  }
+
+  function refreshFile(file) {
+    if (isCssFile(file)) {
+      reloadCssFile(file);
+      return;
+    }
+
+    let block = new Block(file); // minimize DOM queries
+    if (block.valid()) return block.refresh();
+  }
+
+  function setHotReloadCookie(files) {
+    var date = new Date();
+
+    // Hot reload cookie expires in 3 seconds
+    date.setSeconds(date.getSeconds() + 3);
+
+    var elements = files.join(',');
+    var expires = date.toUTCString();
+
+    document.cookie = `hot_reload_files=${elements}; expires=${expires}; path=/`;
+  }
+
+  function refreshPage(files) {
+    setHotReloadCookie(files);
+    console.log('[HotReload] Refreshing entire page');
+    window.location.reload();
+  }
+
+  function handleUpdate(message) {
+    let data = JSON.parse(message.data);
+    var modifiedFiles = data.modified;
+
+    if (isRefreshRequired(modifiedFiles)) {
+      refreshPage(modifiedFiles);
+    } else {
+      modifiedFiles.forEach(refreshFile);
+    }
+  }
+
+  function isCssFile(filename) {
+    return filename.endsWith('.css');
+  }
+
+  function reloadCssFile(filename) {
+    // Find a stylesheet link starting with /assets (locally-served only) containing the filename
+    let links = document.querySelectorAll(`link[href^="/assets"][href*="${filename}"][rel="stylesheet"]`);
+
+    Array.from(links).forEach((link) => {
+      if (!link) {
+        console.log(`[HotReload] Could not find link for stylesheet ${filename}`);
+      } else {
+        console.log(link)
+        link.href = new URL(link.href).pathname + `?v=${Date.now()}`;
+        console.log(`[HotReload] Reloaded stylesheet ${filename}`);
+      }
+    });
+  }
+
+  function isBlockFile(filename) {
+    return new Block(filename).valid();
+  }
+
+  class Block {
+    constructor(filename) {
+      this.filename = filename;
+      this.name = filename.split('/').pop().replace('.liquid', '');
+      const [elements, type] = querySelectHotReloadElements(this.name);
```

Comment:
An idea is to take this out of the constructor so we avoid DOM queries when the file is a snippet

Reply:
This would be an excellent optimization ;)